README.md
lazy_mysql
=======

点击查看[中文版](./README.CH.md)

<br>

TOC
-----

1. [Intro](#intro)
2. [Installation](#installation)
3. [Tutorial](#tutorial)
4. [API](#api)
    1. [Engine](#1-engine)
    2. [Pool](#2-pool)
    3. [Column](#3-column)
    4. [Table](#4-table)

<br>

Intro
------

The module which is based on `MySQL-python` provides four main classes:

* **Engine**   : connect to MySQL server and execute SQL statements.
* **Pool**      : a pool that manages Engine objects.
* **Table**    : a python object mapping table in database.
* **Column** : a python object mapping field in database. 

### Dependencies

* Python 2.6 - 2.7
* MySQLdb-python 1.2.3+

<br>

Installation
--------------

Download from [GitHub](https://github.com/hendiko/PyLazy.git).

    git clone https://github.com/hendiko/PyLazy.git

Or you can simply download `lazy_mysql.py` then put it anywhere in your project.
   
<br>

Tutorial
---------

### 1. Connect to server

Use **Engine**.

    from lazy_mysql import Engine, Pool, Table, Column
    
    engine = Engine('localhost', 'test', 'root', 'root')
   
It had better use **Pool** object to set up a pool to manage **Engine** objects to handle multi threads.

    pool = Pool('localhost', 'test', 'root', 'root', pool_size=4, extras=4)    

### 2. Create **Table** object

To create a **Table** object which maps a table in database, you need to define a class which inherits class **Table**.

    class Schedule(Table):
    
        def __init__(self, table_name='schedule', _engine=engine, *columns):
            super(Schedule, self).__init__(table_name, _engine, *columns)
            
            self.schedule_id = Column('scheduleId')
            self.task_id = Column('taskId')
            self.task_name = Column('taskName')
            self.status = Column('status')

The argument of **_engine** could be either an **Engine** instance or a **Pool** instance.

You could only define those fields you need to use instead of all of the fields that exsits in your database table. A **Column** object needs to be passed in field name to initialize itself, it won't check the field type nor validate the value to be written into database.

### 3. Select

    # Create a Table instance.
    s = Schedule()
          
    # SELECT * FROM schedule LIMIT 1;
    s.select().go() 

    # SELECT * FROM schedule;
    s.select().limit().go()
    
    # SELECT taskName, status FROM schedule WHERE (taskId=1) LIMIT 1;
    s.select(s.task_name, s.status).where(schedule.task_id == 1).go()
    
    # SELECT DISTINCT taskName FROM schedule WHERE (scheduleId=1) GROUP BY taskName ORDER BY taskId DESC LIMIT 1, 4;
    s.select(s.task_name).where(s.schedule_id == 1).distinct().order(s.task_id, desc=True).group_by(s.task_name).limit(1, 4).go()
    
    # SELECT * FROM schedule WHERE (scheduleId=1 AND taskId=2) OR (taskId=2) AND (taskName='query') LIMIT 1;
    s.select().where(s.schedule_id == 1, s.task_id == 3).where(s.task_id == 2).where_and(s.task_name == "query").go()

### 4. Insert

    # INSERT INTO schedule SET taskName='query';
    s.insert(**{s.task_name.name: 'query'}).go()
    
    # or a deprecated way which is more convenient but less reliable.
    s.insert(taskName="query").go()
    

### 5. Update

    # UPDATE schedule SET taskName='query2' WHERE (scheduleId=5) LIMIT 1;
    s.update(**{s.task_name.name: "query2"}).where(s.schedule_id == 5).go()

    # or a deprecated way which is more convenient but less reliable.
    s.update(taskName="query2").where(s.schedule_id == 5).go()
    

### 6. Delete

    # DELETE FROM schedule WHERE (scheduleId=5) LIMIT 1;
    s.delete().where(s.schedule_id == 5).go()
    

### 7. Count

    # SELECT COUNT(DISTINCT scheduleId) AS X FROM schedule WHERE (scheduleId>2) LIMIT 1;
    s.count(s.schedule_id, distinct=True).where(s.schedule_id > 2).go()


API
----

### 1. Engine

**Engine** - connect to MySQL server and execute SQL statements.

#### 1.1. \__init\__(self, host, schema, user, pw, port=3306, charset='utf8', cursor_class='dict', autocommit=True, debug=True, \*args, \*\*kwargs):

* host: host address.
* schema: database name.
* user: user name.
* pw: password.
* port: port.
* charset: charset.
* cursor_class: use DictCursor if 'dict' given, or Cursor.
* autocommit: automatically commit.
* debug: if true, log the every SQL statement executed.
* args: other arguments for MySQLdb.connect()
* kwargs: other keyword arguments for MySQLdb.connect().

#### 1.2. affected_rows

The number of rows that have been affected by executing SQL statement.

#### 1.3. last_executed

The last SQL statement that is executed successfully.

#### 1.4. cursor_class

if a string 'dict' was given, use `DictCursor` instead of `Cursor`.

#### 1.5. connect(self, cursor_class=None)

Return a **Connection** object. If *cursor_class* is *None*, the `self.cursor_class` will be used.

#### 1.6. close(self)

Close connection.

### 2. Pool

#### 2.1. \__init\__(self, host, schema, user, pw, port=3306, charset='utf8', cursor_class='dict', autocommit=True, debug=True, pool_size=2, extras=4, wait_time=5, \*args, \*\*kwargs):

Initialize a **Pool** object to manage a number of **Engine** objects. A queue object that exsits inside the **Pool** object actually reserves all **Engine** objects. You could use the parameter **pool_size** to limit the number of **Engine** in pool. It also allows you to create a number of extra **Engine** objects outside the pool, the extra engines would cost more because they are created on the fly and be destroyed after use, you probably want to use them only in case there are many requests coming suddenly.

The most of arguments of Pool is similar to those in Engine.

* pool_size: the maximum number of engines reserved in pool.
* extras: the maximun number of engines existing outside the pool.
* wait_time: the timeout seconds waiting for Engine to be acquired from the pool.

#### 2.2. count

The current number of Engines both inside or outside the pool.

#### 2.3. put(self, engine=None)

Put the Engine object back to the pool.

#### 2.4. get(self)

Acquire Engine object from the pool.

#### 2.5. spawn_engine(self):

Create a new Engine object.

### 3. Column

#### 3.1. \__init\__(self, name)

The argument name is a string that is name of field in database.

#### 3.2. like(self, other)

The like statement in SQL.

#### 3.3. in_(self, \*other)

The in statement in SQL.

#### 3.4. between(self, floor, ceil)

The between statement in SQL.

#### 3.5. is_null(self, boolean=True)

The IS NULL or IS NOT NULL in SQL.


### 4. Table

#### 4.1. \__init\__(self, table_name, \_engine=None, \*columns)

Initialize a **Table** object.

* table_name: a string that is the name of table.
* \_engine: either an **Engine** object or a **Pool** object.

#### 4.2. add_column(self, \*columns)

Add a column into Table object.

#### 4.3. remove_column(self, \*columns)

Remove a column from Table object.

#### 4.4. binding_engine(self, engine)

Binding either an Engine or a Pool object to this Table object.

#### 4.5. select(self, \*columns)

Execute SELECT statement.

#### 4.6. insert(self, \**assignments)

Execute INSERT statement.

#### 4.7. update(self, \**assignments)

Execute UPDATE statement.

#### 4.8 delete(self)

Execute DELETE statement.

#### 4.9. count(self, column=None, distinct=None)

Execute COUNT statement.

CHANGELOG.md
2015-4-1 version 1.2.6
--------------------------

    1. 增加按多列进行 group by 操作。
    2. Column 对象增加 sum, count, max, min 方法。
    3. 修复 MySQL-python 1.2.3 环境下进行 in 操作时的字符转义问题。

    1. Changed signature of _Select.group_by() to support group by multiple columns.
    2. Added new methods sum, count, max, min to class Column.
    3. Fixed bug over string escaping when using method Column.in_() in MySQL-python 1.2.3.


2015-3-12 version 1.2.1
----------------------------

    1. 增加了 Pool 对象作为连接池。

    1. Added class Pool to manage Engine objects.


2014-4-21 version 1.1.2
----------------------------

    1. _Select类增加group_by()方法。

    1. Added method group_by() in class _Select.


2014-2-26 version 1.1.1
----------------------------

    1. 修改Session执行日志级别，SELECT, COUNT操作属于DEBUG级别，INSERT, UPDATE, DELETE属于INFO级别。

    1. Reset logging level of SELECT, COUNT operations to DEBUG, logging level of INSERT, UPDATE, DELETE operations to INFO.


2014-2-22 version 1.1.0
----------------------------

    1. Column对象增加is_null()方法，支持MYSQL的IS NULL语句。

    1. Improved object Column by adding method is_null() in order to support MySql syntax IS NULL.


2014-1-16 version 1.0.3.0
------------------------------

    1. 修改Table.add_column，Table.remove_column方法签名。

    1. Change signature of Table.add_column, Table.remove_column.


2014-1-16 version 1.0.2.1
------------------------------

    1. 使Table.binding_engine()方法具有返回值。

    1. Making Table.binding_engine() method return itself.


2014-1-16 version 1.0.2.0
------------------------------

    1. 增加SQL执行日志。
    2. 增加Session的affected_rows, last_executed属性。

    1. Added logger, all sql execution would be logged at debug level.
    2. Added property affected_rows, last_executed to _BaseSession and Engine object.


2014-1-13 version 1.0.1.2
------------------------------

    1. 修复Engine.connect()方法。

    1. Fixed bug in Engine.connect().


2014-1-13 version 1.0.1.1
------------------------------

    1. 修改Engine对象接收charset作为初始化参数。

    1. Modified class Engine to accept charset as one of its __init__() arguments.


2014-1-12 version 1.0.1.0
------------------------------

    1. Table.remove_column(self, column) 改进为column参数可以为Column对象。
    2. Engine对象增加若干方法。
    3. 增加clear()方法。
    4. 更名为lazy_mysql。

    1. Improve Table.remove_column to be able to accept Column object.
    2. Add several instance methods to Engine object.
    3. Add method clear() to _BaseSession object.
    4. Renamed module to lazy_mysql.

sample.py
#!/usr/bin/env python
# -*- coding: utf8 -*-

from lazy_mysql import Engine, Table, Column, Pool
import logging

logging.basicConfig(level=logging.DEBUG)

# create engine instance in order to connect local database
engine = Engine('localhost', 'db_site_monitor', 'puppy', 'puppy')

# Using Pool object which is a container to manage multiple Engine instances is a
# better choice when you want to run multi threads with lazy_mysql.
# The argument pool_size means the maximum Engine objects that are saved in pool.
# The argument extras means it would create a number of extra Engine objects if
# there is no one in pool.
pool = Pool('localhost', 'db_site_monitor', 'puppy', 'puppy', pool_size=4, extras=10)


# define Schedule object which is subclass of Table to map table 'tb_schedule_i' in database 'db_site_monitor'
class Schedule(Table):
    def __init__(self, table_name='tb_schedule_i', _engine=engine, *columns):
        """
        This class maps itself to table tb_schedule_i in database db_site_monitor hosted in local machine.
        """
        super(Schedule, self).__init__(table_name, _engine, *columns)

        self.schedule_id = Column('scheduleId')
        self.task_id = Column('taskId')
        self.task_name = Column('taskName')
        self.status = Column('status')
        self.due_datetime = Column('dueDateTime')
        self.start_datetime = Column('startDateTime')
        self.end_datetime = Column('endDateTime')
        self.edit_datetime = Column('editDateTime')


if __name__ == '__main__':
    # SHOW DATABASES
    engine.show_databases()

    # SHOW TABLES
    engine.show_tables()

    # SHOW CREATE TABLE `tb_schedule_i`
    engine.show_create_table('tb_schedule_i')

    # CREATE INSTANCE schedule
    schedule = Schedule()

    # SELECT * FROM tb_schedule_i LIMIT 1
    schedule.select().go()

    # SELECT taskName, status FROM tb_schedule_i WHERE (taskId=1) LIMIT 1
    schedule.select(schedule.task_name, schedule.status).where(schedule.task_id == 1).go()

    # SELECT DISTINCT * FROM tb_schedule_i ORDER BY dueDateTime DESC LIMIT 1, 2
    schedule.select().distinct().order(schedule.due_datetime, desc=True).limit(1, 2).go()

    # UPDATE tb_schedule_i SET status=0 WHERE (scheduleId=2 AND taskId=1) LIMIT 1
    schedule.update(**{schedule.status.name: 0}).where(schedule.schedule_id == 2, schedule.task_id == 1).go()

    # INSERT INTO tb_schedule_i SET status=1
    new_schedule_id = schedule.insert(**{schedule.status.name: 1}).go()

    # DELETE FROM tb_schedule_i WHERE (scheduleId=93) LIMIT 1
    schedule.delete().where(schedule.schedule_id == new_schedule_id).go()

    # SELECT COUNT(DISTINCT scheduleId) AS X FROM tb_schedule_i WHERE (scheduleId>50) LIMIT 1
    schedule.count(schedule.schedule_id, distinct=True).where(schedule.schedule_id > 50).go()

    pass

lazy_mysql.py
# !/usr/bin/env python
# -*- coding: utf-8 -*-
# Author : Xavier Yin
# Date    : 2013-12-26

"""
The lazy_mysql is a simple and light ORM framework based on MySQLdb.

The module provides five main classes which are Engine, Pool, Session, Table, Column.
It provides five SQL operations, they are SELECT, INSERT, DELETE, UPDATE, COUNT.

本模块包括 Engine, Pool, Session, Table, Column 五个对象。
目前支持 SELECT, INSERT, DELETE, UPDATE, COUNT 操作。
条件支持 WHERE, ORDER BY, GROUP BY, DISTINCT, LIMIT。
"""
__author__ = 'Xavier Yin'
__version__ = '1.2.6'
__date__ = '2015-4-1'


from datetime import datetime
from MySQLdb import cursors
from Queue import Queue, Full, Empty
from threading import RLock
import logging
import MySQLdb

MySQLdb.threadsafety = 1
logger = logging.getLogger(__name__)


class Engine(object):
    """The engine to connect database."""
    def __init__(self, host, schema, user, pw, port=3306, charset='utf8',
                 cursor_class='dict', autocommit=True, debug=True, *args, **kwargs):
        """初始化数据库连接参数。
        :param host: 数据库主机。
        :param schema: 数据库名称。
        :param user: 用户名。
        :param pw: 密码。
        :param port: 端口。
        :param charset: 字符集。
        :param cursor_class: 游标类，默认值为 dict，其他为 tuple。
        :param autocommit: 自动提交。
        :param debug: 调试模式。
        :param args: 其他参数。
        :param kwargs: 其他参数。
        """
        self.host = host
        self.schema = schema
        self.user = user
        self.pw = pw
        self.port = port
        self.connection = None
        self.charset = charset
        self._cursor_class = cursor_class
        self.autocommit = autocommit
        self.debug = debug
        self.args = args
        self.kwargs = kwargs
        self.affected_rows = 0
        self.last_executed = ''

    @property
    def cursor_class(self):
        if self._cursor_class == 'dict':
            return cursors.DictCursor
        else:
            return cursors.Cursor

    @cursor_class.setter
    def cursor_class(self, _class):
        self._cursor_class = _class

    def connect(self, cursor_class=None):
        """建立数据库连接。"""
        if not self.connection:
            self.connection = MySQLdb.connect(
                self.host, self.user, self.pw, self.schema, port=self.port,
                charset=self.charset, cursorclass=cursor_class)
            self.connection.autocommit(self.autocommit)
        if cursor_class is None:
            self.connection.cursorclass = self.cursor_class
        elif cursor_class == "dict":
            self.connection.cursorclass = cursors.DictCursor
        else:
            self.connection.cursorclass = cursors.Cursor
        return self.connection

    def close(self):
        try:
            self.connection.close()
        except Exception as e:
            if self.debug:
                logger.exception(str(e))
        finally:
            self.connection = None

    def create_database(self, table_name, confirm=False):
        """创建新的数据库。
        :param table_name: 数据表名称。
        :param confirm: 如果为真，则新建数据表。
        """
        if confirm:
            sql = 'CREATE DATABASE `%(name)s`' % {'name': table_name}
            return self._transaction(sql)

    def drop_database(self, table_name, confirm=False):
        """
        丢弃数据库
        :param table_name: 数据库名称。
        :param confirm: 安全确认。仅当confirm为True时才执行丢弃操作。
        """
        if confirm is True:
            sql = 'DROP DATABASE `%(name)s`' % {'name': table_name}
            return self._transaction(sql)

    def show_create_table(self, table_name):
        """显示数据库创建SQL语句。"""
        sql = 'SHOW CREATE TABLE `%(name)s`' % {'name': table_name}
        return self._transaction(sql, True)[0]['Create Table']

    def show_databases(self):
        """显示所有数据库名称。"""
        sql = 'SHOW DATABASES'
        return [d['Database'] for d in self._transaction(sql, True)]

    def show_tables(self):
        """显示所有数据表名称。"""
        sql = 'SHOW TABLES'
        return [d['Tables_in_%s' % self.schema] for d in self._transaction(sql, True)]

    def set_names(self, charset='utf8'):
        """设置输出字符集编码。"""
        sql = 'SET NAMES `%(name)s`' % {'name': charset.upper()}
        return self._transaction(sql)

    def _transaction(self, sql, fetch=False, cursor_class=None):
        """不要直接调用本方法，执行SQL语句并返回结果。"""
        self.affected_rows, self.last_executed, timestamp = 0, '', datetime.now().strftime('%Y-%m-%d %H:%M:%S')
        try:
            conn = self.connect(cursor_class)
            cursor = conn.cursor()
            self.affected_rows = cursor.execute(sql)
        except Exception as e:
            if self.debug:
                logger.exception(str(e))
            self.close()
            conn = self.connect(cursor_class)
            cursor = conn.cursor()
            self.affected_rows = cursor.execute(sql)

        self.last_executed = cursor._last_executed
        if self.debug:
            logger.debug('%s : %s ROW(S) AFFECTED WITH SQL: %s', timestamp, self.affected_rows, self.last_executed)
        result = cursor.fetchall() if fetch else self.affected_rows
        cursor.close()
        return result


class Pool(object):

    def __init__(self, host, schema, user, pw, port=3306, charset='utf8',
                 cursor_class='dict', autocommit=True, debug=True, pool_size=2,
                 extras=4, wait_time=5, *args, **kwargs):
        """初始化数据库连接参数。
        :param host: 数据库主机。
        :param schema: 数据库。
        :param user: 用户名。
        :param pw: 密码。
        :param port: 端口。
        :param charset: 字符集。
        :param cursor_class: 游标类型。
        :param autocommit: 自动提交。
        :param debug: 调试模式。
        :param pool_size: 连接池大小。
        :param extras: 允许超出连接池大小。
        :param wait_time: 从连接池获取 Engine 对象超时时间。
        :param args: 其他参数。
        :param kwargs: 其他参数。
        """
        self.host = host
        self.schema = schema
        self.user = user
        self.pw = pw
        self.port = port
        self.connection = None
        self.charset = charset
        self.cursor_class = cursor_class
        self.autocommit = autocommit
        self.debug = debug
        self.args = args
        self.kwargs = kwargs
        self.affected_rows, self.last_executed = 0, ''
        self.wait_time = wait_time
        self.lock = RLock()
        self.pool_size = pool_size
        self.limits = extras + self.pool_size
        self._count = 0
        self.pool = Queue(self.pool_size)

    @property
    def count(self):
        return self._count

    def spawn_engine(self):
        self.lock.acquire()
        if self._count < self.limits:
            self._count += 1
            _engine = Engine(
                self.host, self.schema, self.user, self.pw, self.port, self.charset,
                self.cursor_class, self.autocommit, self.debug, *self.args, **self.kwargs)
            self.lock.release()
        else:
            self.lock.release()
            _engine = self.get()

        return _engine

    def put(self, engine=None):
        if not isinstance(engine, Engine):
            engine = self.spawn_engine()
        try:
            self.pool.put(engine, False)
        except Full:
            self.lock.acquire()
            self._count -= 1
            self.lock.release()

    def get(self):
        try:
            engine = self.pool.get(timeout=self.wait_time)
            self.pool.task_done()
            return engine
        except Empty:
            self.lock.acquire()
            if self._count < self.limits:
                self.lock.release()
                return self.spawn_engine()
            else:
                self.lock.release()
                engine = self.pool.get()
                self.pool.task_done()
                return engine


class Column(object):
    def __init__(self, name):
        """数据库字段。"""
        self.name = name

    def __str__(self):
        return self.name

    def __eq__(self, other):
        name = '{0}_{1}'.format(self.name, 'eq')
        return '{0}=%({1})s'.format(self.name, name), {name: other}

    def __ne__(self, other):
        name = '{0}_{1}'.format(self.name, 'ne')
        return '{0}<>%({1})s'.format(self.name, name), {name: other}

    def __gt__(self, other):
        name = '{0}_{1}'.format(self.name, 'gt')
        return '{0}>%({1})s'.format(self.name, name), {name: other}

    def __ge__(self, other):
        name = '{0}_{1}'.format(self.name, 'ge')
        return '{0}>=%({1})s'.format(self.name, name), {name: other}

    def __lt__(self, other):
        name = '{0}_{1}'.format(self.name, 'lt')
        return '{0}<%({1})s'.format(self.name, name), {name: other}

    def __le__(self, other):
        name = '{0}_{1}'.format(self.name, 'le')
        return '{0}<=%({1})s'.format(self.name, name), {name: other}

    def like(self, other):
        name = '{0}_{1}'.format(self.name, 'like')
        return '{0} LIKE %({1})s'.format(self.name, name), {name: other}

    def in_(self, *other):
        names, maps = [], {}
        for index in range(len(other)):
            name = '{0}_{1}_{2}'.format(self.name, 'in', index)
            names.append("%({0})s".format(name))
            maps[name] = other[index]
        return '{0} IN ({1})'.format(self.name, ",".join(names)), maps

    def not_in(self, *other):
        names, maps = [], {}
        for index in range(len(other)):
            name = '{0}_{1}_{2}'.format(self.name, 'in', index)
            names.append("%({0})s".format(name))
            maps[name] = other[index]
        return '{0} NOT IN ({1})'.format(self.name, ",".join(names)), maps

    def between(self, floor, ceil):
        _floor = '{0}_{1}'.format(self.name, 'floor')
        _ceil = '{0}_{1}'.format(self.name, 'ceil')
        return '{0} BETWEEN %({1})s AND %({2})s'.format(self.name, _floor, _ceil), {_floor: floor, _ceil: ceil}

    def is_null(self, boolean=True):
        name = '{0}_{1}'.format(self.name, 'is_null')
        if boolean:
            return '{0} IS %({1})s'.format(self.name, name), {name: None}
        else:
            return '{0} IS NOT %({1})s'.format(self.name, name), {name: None}

    def sum(self):
        return 'SUM({0})'.format(self.name)

    def count(self):
        return "COUNT({0})".format(self.name)

    def max(self):
        return "MAX({0})".format(self.name)

    def min(self):
        return "MIN({0})".format(self.name)


class Table(object):
    """数据表对象。"""

    def __init__(self, table_name, _engine=None, *columns):
        """初始化数据表，可以在columns参数中动态传入字段名称，或者覆写本方法并在新方法中直接定义字段。"""
        self.engine = _engine
        self.table_name = table_name
        self.add_column(*columns)

    def add_column(self, *columns):
        """增加字段。"""
        for column in columns:
            setattr(self, column.name, column) if isinstance(column, Column) else setattr(self, column, Column(column))
        return self

    def remove_column(self, *columns):
        """删除字段。"""
        for column in columns:
            self.__delattr__(column.name) if isinstance(column, Column) else self.__delattr__(column)
        return self

    def binding_engine(self, engine):
        """绑定用来建立数据库连接的Engine对象。"""
        self.engine = engine
        return self

    def select(self, *columns):
        """SELECT操作。"""
        return _Select(self.engine, self.table_name, 'SELECT', *columns)

    def insert(self, **assignments):
        """INSERT操作。"""
        return _Insert(self.engine, self.table_name, 'INSERT', **assignments)

    def update(self, **assignments):
        """UPDATE操作。"""
        return _Update(self.engine, self.table_name, 'UPDATE', **assignments)

    def delete(self):
        """DELETE操作。"""
        return _Delete(self.engine, self.table_name, 'DELETE')

    def count(self, column=None, distinct=None):
        """COUNT操作。"""
        return _Count(self.engine, self.table_name, 'COUNT', column, distinct)


class _BaseSession(object):
    def __init__(self, engine, table_name, action, *columns, **assignments):
        self.engine, self.table_name, self.action = engine, table_name, action
        self._columns, self._assignments = columns, assignments

        self._where_clause, self._where_dict = '', {}
        self._distinct_clause = self._order_clause = self._limit_clause = ''
        self.affected_rows, self.last_executed = 0, ''
        self.limit(1)   # for safety consideration

    def clear(self):
        self._where_clause, self._where_dict = '', {}
        self._distinct_clause = self._order_clause = self._limit_clause = ''
        self.limit(1)   # for safety consideration
        return self

    def distinct(self, flag=True):
        """增加DISTINCT条件。"""
        self._distinct_clause = 'DISTINCT' if flag else ''
        return self

    def where(self, *columns):
        """
        执行WHERE语句。
        :param columns: Column对象。
        """
        if columns:
            if self._where_clause:
                new_clause = '({0})'.format(' AND '.join([col[0] for col in columns]))
                self._where_clause = ' OR '.join([self._where_clause, new_clause])
            else:
                self._where_clause = 'WHERE ({0})'.format(' AND '.join([col[0] for col in columns]))
            self._where_dict.update(reduce(lambda x, y: x + y, [column[1].items() for column in columns]))
        else:
            self._where_clause, self._where_dict = '', {}
        return self

    def where_and(self, *columns):
        if columns:
            if self._where_clause:
                new_clause = '({0})'.format(' AND '.join([col[0] for col in columns]))
                self._where_clause = ' AND '.join([self._where_clause, new_clause])
            else:
                self._where_clause = 'WHERE ({0})'.format(' AND '.join([col[0] for col in columns]))
            self._where_dict.update(reduce(lambda x, y: x + y, [column[1].items() for column in columns]))
        else:
            self._where_clause, self._where_dict = '', {}
        return self

    def order(self, column=None, desc=False):
        """执行ORDER条件。"""
        self._order_clause = ('ORDER BY {0} {1}'.format(str(column), 'DESC' if desc else 'ASC')) if column else ''
        return self

    def limit(self, number=None, step=None):
        """执行LIMIT条件。"""
        self._limit_clause = '' if number is None else 'LIMIT %s' % number
        if None not in (number, step):
            self._limit_clause = '%s, %s' % (self._limit_clause, step)
        return self

    def _transaction(self, clauses, sql_dict, cursor_class):
        sql_clause = ' '.join([clause for clause in clauses if clause])
        self.affected_rows, self.last_executed, timestamp = 0, '', datetime.now().strftime('%Y-%m-%d %H:%M:%S')

        _engine = self.engine.get() if isinstance(self.engine, Pool) else self.engine

        try:
            conn = _engine.connect(cursor_class)
            cursor = conn.cursor()
            self.affected_rows = cursor.execute(sql_clause, sql_dict)
        except Exception as e:
            if _engine.debug:
                logger.exception(str(e))
            _engine.close()
            conn = _engine.connect(cursor_class)
            cursor = conn.cursor()
            self.affected_rows = cursor.execute(sql_clause, sql_dict)
        finally:
            if isinstance(self.engine, Pool):
                self.engine.put(_engine)

        self.last_executed = cursor._last_executed
        if _engine.debug:
            logger.debug('%s: %s ROW(S) AFFECTED WITH SQL: %s', timestamp, self.affected_rows, self.last_executed)
        result = None
        if self.action == 'SELECT':
            result = cursor.fetchall()
        elif self.action == 'INSERT':
            result = cursor.lastrowid
        elif self.action in ('UPDATE', 'DELETE'):
            result = self.affected_rows
        elif self.action == 'COUNT':
            result = cursor.fetchall()[0]['X']
        cursor.close()
        return result


class _Select(_BaseSession):
    def __init__(self, engine, table_name, action, *columns, **assignments):
        super(_Select, self).__init__(engine, table_name, action, *columns, **assignments)
        self._action_clause = ', '.join([str(column) for column in self._columns]) or '*'
        self._group_by_clause = None

    def group_by(self, *columns):
        self._group_by_clause = 'GROUP BY %s' % (','.join([str(column) for column in columns]))
        return self

    def go(self, cursor_class=None):
        clauses = [self.action, self._distinct_clause, self._action_clause, 'FROM', self.table_name,
                   self._where_clause, self._group_by_clause, self._order_clause, self._limit_clause]
        sql_dict = self._where_dict
        return self._transaction(clauses, sql_dict, cursor_class)


class _Insert(_BaseSession):
    def __init__(self, engine, table_name, action, *columns, **assignments):
        super(_Insert, self).__init__(engine, table_name, action, *columns, **assignments)
        self._action_clause = ', '.join(['{0}=%(_insert_{0})s'.format(k) for k in self._assignments])
        self._action_dict = dict(('_insert_%s' % k, v) for k, v in self._assignments.items())

    def go(self, cursor_class=None):
        clauses = [self.action, 'INTO', self.table_name, 'SET', self._action_clause]
        sql_dict = self._action_dict
        return self._transaction(clauses, sql_dict, cursor_class)


class _Update(_BaseSession):
    def __init__(self, engine, table_name, action, *columns, **assignments):
        super(_Update, self).__init__(engine, table_name, action, *columns, **assignments)
        self._action_clause = ', '.join(['{0}=%(_update_{0})s'.format(k) for k in self._assignments])
        self._action_dict = dict(('_update_%s' % k, v) for k, v in self._assignments.items())

    def go(self, cursor_class=None):
        clauses = [self.action, self.table_name, 'SET', self._action_clause, self._where_clause,
                   self._order_clause, self._limit_clause]
        sql_dict = dict()
        sql_dict.update(self._action_dict, **self._where_dict)
        return self._transaction(clauses, sql_dict, cursor_class)


class _Delete(_BaseSession):
    def __init__(self, engine, table_name, action, *columns, **assignments):
        super(_Delete, self).__init__(engine, table_name, action, *columns, **assignments)

    def go(self, cursor_class=None):
        clauses = [self.action, 'FROM', self.table_name, self._where_clause, self._order_clause, self._limit_clause]
        sql_dict = self._where_dict
        return self._transaction(clauses, sql_dict, cursor_class)


class _Count(_BaseSession):
    def __init__(self, engine, table_name, action, *columns, **assignments):
        super(_Count, self).__init__(engine, table_name, action, *columns, **assignments)
        col, distinct = self._columns
        self._action_clause = ' '.join(['DISTINCT' if col and distinct else '', "'*'" if col is None else str(col)])
        self._action_clause = 'COUNT({0})'.format(self._action_clause)

    def go(self, cursor_class=None):
        clauses = ['SELECT', self._action_clause, 'AS X FROM', self.table_name, self._where_clause, self._limit_clause]
        sql_dict = self._where_dict
        return self._transaction(clauses, sql_dict, cursor_class)


if __name__ == '__main__':
    pass